### PR TITLE
database arguments

### PR DESCRIPTION
**Pull request by red-avtovo**
Added posibility to setup connection with databases, which requires some extra arguments in URL
For example:
```
postgresql://postgres@localhost:5432/diesel_example?ssl=true
```

### CODE REVIEW

**Comment by red-avtovo** [comment]
Anyone?

**Comment by killercup** [comment]
Hey, sorry for the lack of feedback here! I'm at a conference currently,
but maybe @Eijebong or @sgrif can have a look?


Viele Grüße,
Pascal

**Comment by sgrif** [comment]
As per [the docs](http://docs.diesel.rs/diesel/pg/struct.PgConnection.html)

> The connection string expected by PgConnection::establish should be a PostgreSQL connection string, as documented at http://www.postgresql.org/docs/9.4/static/libpq-connect.html#LIBPQ-CONNSTRING

All available arguments are documented at http://www.postgresql.org/docs/9.4/static/libpq-connect.html#LIBPQ-CONNSTRING

**Comment by sgrif** [comment]
Sorry I haven't had my coffee yet I thought this was an issue not a PR. >_>

File: diesel_cli/src/database.rs [hunk 1]
```diff
@@ -316,8 +316,16 @@ pub fn database_url(matches: &ArgMatches) -> String {
 #[cfg(any(feature="postgres", feature="mysql"))]
 fn change_database_of_url(database_url: &str, default_database: &str) -> (String, String) {
     let mut split: Vec<&str> = database_url.split('/').collect();
-    let database = split.pop().unwrap();
-    let new_url = format!("{}/{}", split.join("/"), default_database);
+    let database_name_with_arguments: Vec<&str> = split.pop().unwrap().split('?').collect();
```
Comment (sgrif) [inline diesel_cli/src/database.rs]: Can we use the URL crate for this rather than doing our own parsing?

Comment (YetAnotherMinion) [reply diesel_cli/src/database.rs]: @sgrif You mean code looking like this PR? https://github.com/diesel-rs/diesel/pull/861/files

**Comment by sgrif** [comment]
Closing due to inactivity. Feel free to open a new PR if you're interested in continuing to work on this. For this to be merged, I'd like to see the URL crate used rather than manual parsing.

**Comment by YetAnotherMinion** [comment]
@sgrif 
```diff
diff --git a/diesel_cli/Cargo.toml b/diesel_cli/Cargo.toml
index 990e6dd..ff855e5 100644
--- a/diesel_cli/Cargo.toml
+++ b/diesel_cli/Cargo.toml
@@ -22,6 +22,7 @@ dotenv = ">=0.8, <0.11"
 infer_schema_internals = { version = "1.1.0" }
 clippy = { optional = true, version = "=0.0.174" }
 migrations_internals = { version = "1.1.0" }
+url = "1.6"
 
 [dev-dependencies]
 difference = "1.0"
diff --git a/diesel_cli/src/database.rs b/diesel_cli/src/database.rs
index 5576ff2..1c0a6ce 100644
--- a/diesel_cli/src/database.rs
+++ b/diesel_cli/src/database.rs
@@ -5,6 +5,7 @@ use diesel::*;
 use migrations_internals as migrations;
 #[cfg(any(feature = "postgres", feature = "mysql"))]
 use super::query_helper;
+use url::Url;
 
 use database_error::{DatabaseError, DatabaseResult};
 
@@ -323,10 +324,20 @@ pub fn database_url(matches: &ArgMatches) -> String {
 
 #[cfg(any(feature = "postgres", feature = "mysql"))]
 fn change_database_of_url(database_url: &str, default_database: &str) -> (String, String) {
-    let mut split: Vec<&str> = database_url.split('/').collect();
-    let database = split.pop().unwrap();
-    let new_url = format!("{}/{}", split.join("/"), default_database);
-    (database.to_owned(), new_url)
+    // This method accepts a MySQL or Postgres connection string that conforms to URL BNF grammar
+    // only (and not the database connection string grammars) and returns the name of the database
+    // and a new connection string to the database specified by the default_database argument. If
+    // the database name is not specified in the connection string then we default to an empty
+    // string. NOTE: This function does not validate connection strings (look at the signature, no
+    // Option or Result is there).
+
+    // TODO: handle unix connection strings with percent encoded paths such as:
+    //  postgres://%2Fvar%2Flib/dbname
+    let mut url = Url::parse(database_url).expect("failed to parse connection string");
+    let database: String = url.path().trim_left_matches('/').to_string();
+    url.set_path(default_database);
+
+    (database, url.into_string())
 }
 
 #[cfg_attr(feature = "clippy", allow(needless_pass_by_value))]
@@ -341,14 +352,42 @@ mod tests {
 
     #[test]
     fn split_pg_connection_string_returns_postgres_url_and_database() {
-        let database = "database".to_owned();
-        let base_url = "postgresql://localhost:5432".to_owned();
-        let database_url = format!("{}/{}", base_url, database);
-        let postgres_url = format!("{}/{}", base_url, "postgres");
-        assert_eq!(
-            (database, postgres_url),
-            change_database_of_url(&database_url, "postgres")
-        );
+        // format: (original_string, dbname, expected_changed_string)
+        let test_cases: [(&'static str, &'static str, &'static str); 10] = [
+            ("postgresql://", "", "postgresql:///postgres"),
+            ("postgresql://localhost", "" ,"postgresql://localhost/postgres"),
+            ("postgresql://localhost:5433", "", "postgresql://localhost:5433/postgres"),
+            ("postgresql://localhost/mydb", "mydb", "postgresql://localhost/postgres"),
+            ("postgresql://user@localhost", "","postgresql://user@localhost/postgres"),
+            ("postgresql://user:secret@localhost", "", "postgresql://user:secret@localhost/postgres"),
+            ("postgresql://other@localhost/otherdb?connect_timeout=10&application_name=myapp", "otherdb", "postgresql://other@localhost/postgres?connect_timeout=10&application_name=myapp" ),
+            ("postgresql:///mydb?host=localhost&port=5433", "mydb", "postgresql:///postgres?host=localhost&port=5433"),
+            ("postgresql://[2001:db8::1234]/database", "database", "postgresql://[2001:db8::1234]/postgres"),
+            ("postgresql:///dbname?host=/var/lib/postgresql", "dbname", "postgresql:///postgres?host=/var/lib/postgresql"),
+            // TODO: Support opaque origins
+            //("postgresql://%2Fvar%2Flib%2Fpostgresql/dbname", "dbname", "postgresql://%2Fvar%2Flib%2Fpostgresql/postgres"),
+        ];
+        for &(database_url, database, postgres_url) in &test_cases {
+            assert_eq!((database.to_owned(), postgres_url.to_owned()), change_database_of_url(database_url, "postgres"));
+        }
+    }
+
+    #[test]
+    fn split_mysql_connection_string_returns_mysql_url_and_database() {
+        // format: (original_string, dbname, expected_changed_string)
+        let test_cases: [(&'static str, &'static str, &'static str); 8] = [
+            ("mysql://", "", "mysql:///information_schema"),
+            ("mysql://localhost", "" ,"mysql://localhost/information_schema"),
+            ("mysql://localhost:5433", "", "mysql://localhost:5433/information_schema"),
+            ("mysql://localhost/mydb", "mydb", "mysql://localhost/information_schema"),
+            ("mysql://user@localhost", "","mysql://user@localhost/information_schema"),
+            ("mysql://user:secret@localhost", "", "mysql://user:secret@localhost/information_schema"),
+            ("mysql://other@localhost/otherdb", "otherdb", "mysql://other@localhost/information_schema"),
+            ("mysql://other@localhost:1122/otherdb", "otherdb", "mysql://other@localhost:1122/information_schema"),
+        ];
+        for &(database_url, database, mysql_url) in &test_cases {
+            assert_eq!((database.to_owned(), mysql_url.to_owned()), change_database_of_url(database_url, "information_schema"));
+        }
     }
 
     #[test]
diff --git a/diesel_cli/src/main.rs b/diesel_cli/src/main.rs
index 3f689b1..4025e51 100644
--- a/diesel_cli/src/main.rs
+++ b/diesel_cli/src/main.rs
@@ -15,6 +15,7 @@ extern crate clap;
 #[cfg_attr(any(feature = "mysql", feature = "postgres"), macro_use)]
 extern crate diesel;
 extern crate dotenv;
+extern crate url;
 extern crate infer_schema_internals;
 extern crate migrations_internals;
 ```

**Comment by sgrif** [comment]
Feel free to open a PR. We should open a PR to the URL crate to fix the percent encoded case as well.

**Pull request closed without merging**